File: templates/matlab/pacf_plot.m
function fig = pacf_plot()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(2);
    nlags = 20; n = 400;
    y = zeros(1, n);
    for i = 3:n
        y(i) = 0.6*y(i-1) - 0.3*y(i-2) + randn;   % AR(2) demo series
    end
    pacf = pacf_(y, nlags);
    ci = 1.96/sqrt(n);
    lags = 1:nlags;
    fig = figure; hold on;
    hci = fill([0 nlags+1 nlags+1 0], [-ci -ci ci ci], palette('cat',1), ...
               'FaceAlpha', 0.15, 'EdgeColor', 'none');
    plot([0 nlags+1], [0 0], 'Color', [0.4 0.4 0.4], 'LineWidth', 0.8);
    xs = [lags; lags; nan(1, nlags)];
    ys = [zeros(1, nlags); pacf; nan(1, nlags)];
    plot(xs(:), ys(:), 'Color', palette('cat',1), 'LineWidth', 1.5);
    plot(lags, pacf, 'o', 'Color', palette('cat',2), 'MarkerSize', 4, ...
         'MarkerFaceColor', palette('cat',2), 'LineStyle', 'none');
    xlim([0 nlags+1]);
    xlabel('lag'); ylabel('partial autocorrelation');
    title('Partial autocorrelation (AR(2) demo)');
    legend(hci, {'95% confidence band'});
    grid on;
end

function pacf = pacf_(y, nlags)
    % Durbin-Levinson recursion on the sample autocorrelation
    y = y - mean(y);
    n = numel(y);
    r = zeros(1, nlags + 1);
    for k = 0:nlags
        r(k+1) = sum(y(1:n-k).*y(1+k:n));
    end
    r = r / r(1);
    phi = zeros(nlags + 1, nlags + 1);        % phi(k+1, j+1) ~ phi_{k,j}
    phi(2, 2) = r(2);
    for k = 2:nlags
        num = r(k+1) - phi(k, 2:k)*fliplr(r(2:k))';
        den = 1 - phi(k, 2:k)*r(2:k)';
        phi(k+1, k+1) = num/den;
        phi(k+1, 2:k) = phi(k, 2:k) - phi(k+1, k+1)*fliplr(phi(k, 2:k));
    end
    pacf = diag(phi)';
    pacf = pacf(2:end);
end

File: templates/python/pacf_plot.py
"""pacf_plot: 偏自相关图（AR(2) 合成序列，Durbin-Levinson 求 PACF + 置信带）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def _pacf(y, nlags):
    y = y - y.mean()
    r = np.array([np.dot(y[:len(y)-k], y[k:]) for k in range(nlags + 1)])
    r = r / r[0]
    phi = np.zeros((nlags + 1, nlags + 1))
    phi[1, 1] = r[1]
    for k in range(2, nlags + 1):
        num = r[k] - np.dot(phi[k-1, 1:k], r[1:k][::-1])
        den = 1 - np.dot(phi[k-1, 1:k], r[1:k])
        phi[k, k] = num / den
        phi[k, 1:k] = phi[k-1, 1:k] - phi[k, k] * phi[k-1, 1:k][::-1]
    return np.array([phi[k, k] for k in range(1, nlags + 1)])

def make_figure(nlags=20, title='Partial autocorrelation (AR(2) demo)'):
    apply_theme()
    rng = np.random.default_rng(2)
    n = 400
    y = np.zeros(n)
    for i in range(2, n):
        y[i] = 0.6 * y[i-1] - 0.3 * y[i-2] + rng.standard_normal()
    pacf = _pacf(y, nlags)
    ci = 1.96 / np.sqrt(n)
    lags = np.arange(1, nlags + 1)
    fig, ax = plt.subplots()
    ax.axhspan(-ci, ci, color=cycle(0), alpha=0.15,
               label='95% confidence band')
    ax.axhline(0, color='#666666', linewidth=0.8)
    ax.vlines(lags, 0, pacf, color=cycle(0), linewidth=1.5)
    ax.plot(lags, pacf, 'o', color=cycle(1), markersize=4)
    ax.set_xlabel('lag'); ax.set_ylabel('partial autocorrelation')
    ax.set_title(title)
    ax.legend(frameon=False)
    ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)
